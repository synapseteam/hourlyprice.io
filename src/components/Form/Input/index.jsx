import React from "react";
import PropTypes from "prop-types";

import { errorsMessages } from "configure";

export default function Input({
  labelName,
  register,
  placeholder,
  changeHandler,
  inputName,
  errors,
}) {
  return (
    <>
      <label>
        {labelName}
        <input
          {...register(inputName)}
          type="text"
          placeholder={placeholder}
          autoComplete="off"
          onChange={changeHandler}
        />
        {errors[inputName] && (
          <p className="error-text">{errorsMessages[inputName]}</p>
        )}
      </label>
    </>
  );
}

Input.propTypes = {
  labelName: PropTypes.string,
  register: PropTypes.func,
  placeholder: PropTypes.string,
  changeHandler: PropTypes.func,
  inputName: PropTypes.string,
  errors: PropTypes.object,
};
