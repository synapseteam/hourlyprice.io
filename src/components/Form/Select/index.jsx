import React from "react";
import uniqid from "uniqid";
import PropTypes from "prop-types";

import { errorsMessages } from "configure";

export default function Select({
  labelName,
  inputName,
  register,
  changeHandler,
  value,
  optionsArr,
  errors,
}) {
  return (
    <>
      <label>
        {labelName}:
        <select {...register(inputName)} onChange={changeHandler} value={value}>
          {optionsArr &&
            optionsArr.map((name) => (
              <option key={uniqid()} value={name}>
                {name}
              </option>
            ))}
        </select>
        {errors[inputName] && <p>{errorsMessages[inputName]}</p>}
      </label>
    </>
  );
}

Select.propTypes = {
  labelName: PropTypes.string,
  register: PropTypes.func,
  placeholder: PropTypes.string,
  changeHandler: PropTypes.func,
  inputName: PropTypes.string,
  errors: PropTypes.object,
  value: PropTypes.string,
  optionsArr: PropTypes.arrayOf(PropTypes.string),
};
