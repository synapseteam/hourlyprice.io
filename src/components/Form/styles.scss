$purple: #24274a;
$dark-purple: #171d3d;
$light-purple: #31335a;
$blue: #5599ec;
$white: #ffffff;

.form-container {
  display: flex;

  form {
    display: flex;
    flex-direction: column;
    width: 100%;

    label {
      display: flex;
      flex-direction: column;
      font-size: 0.7rem;
      color: $blue;
      margin-top: 1.6rem;

      &:first-child {
        margin-top: 0;
      }

      input,
      select {
        background-color: $dark-purple;
        border: 1px solid $light-purple;
        border-radius: 0.3rem;
        color: $white;
        padding: 0.8rem 1rem;
        margin-top: 0.4rem;
      }

      select {
        -moz-appearance: none; /* Firefox */
        -webkit-appearance: none; /* Safari and Chrome */
        appearance: none;
      }
    }
  }
}
