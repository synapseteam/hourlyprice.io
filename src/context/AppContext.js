import { createContext, useReducer, useContext } from "react";

const AppThemeContext = createContext();

const initState = {
  darkMode: true,
};

const TOGGLE_THEME = "TOGGLE_THEME";

function reducer(state, action) {
  switch (action.type) {
    case TOGGLE_THEME:
      return {
        ...state,
        darkMode: !state.darkMode,
      };
    default:
      throw Error("impossible case while switching Theme");
  }
}

export const AppThemeProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initState);
  const value = [state, dispatch];

  return (
    <AppThemeContext.Provider value={value}>
      {children}
    </AppThemeContext.Provider>
  );
};

export const useAppThemeContext = () => {
  const context = useContext(AppThemeContext);

  if (!context) {
    throw Error("useAppThemeContext should be used within AppThemeProvider");
  }

  return context;
};

export const toggleTheme = () => {
  return { type: TOGGLE_THEME, payload: null };
};

export default AppThemeContext;
