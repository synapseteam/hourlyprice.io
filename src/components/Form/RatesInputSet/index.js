import React, { useEffect } from "react";
import { useSelector } from "react-redux";
import uniqid from "uniqid";

import "components/Form/RatesInputSet/styles.scss";

export default function RatesInputSet({ register, chosenCurrency }) {
  const allCurrenciesArr = useSelector((state) => state.rates.allCurrencies);

  const allCurrenciesNames = allCurrenciesArr.map((el) => ({
    ...el,
    basic: el.name === chosenCurrency,
  }));

  return (
    <div className="rates-input-set">
      {allCurrenciesNames.map((el) => {
        return (
          <input
            {...register(el.name)}
            key={uniqid()}
            className="rates-input"
            type="text"
            autoComplete="off"
            placeholder={el.name}
          />
        );
      })}
    </div>
  );
}
