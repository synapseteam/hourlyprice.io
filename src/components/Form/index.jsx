import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";

import Input from "components/Form/Input";
import Select from "components/Form/Select";
import RatesInputSet from "components/Form/RatesInputSet";
import { submitFieldsData, setRequestErr } from "store/mainReducer";
import {
  getNewRatesThunkCreator,
  setManualRates,
} from "store/currenciesReducer";
import { formSchema, ratesSources } from "configure";
import {
  convertStrTimeToNum,
  handleTimeChange,
  handlePriceChange,
} from "utils";
import { ratesUpdatingTimeFrame } from "configure";
import { useAppThemeContext } from "context/AppContext";
import { useCustomTranslation } from "i18n";

import "components/Form/styles.scss";

export default function Form() {
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm({
    resolver: yupResolver(formSchema),
  });

  const [state] = useAppThemeContext();

  const [t] = useCustomTranslation();

  const allCurrencies = useSelector((state) => state.rates.allCurrencies);
  const ratesSource = useSelector((state) => state.rates.ratesSource);
  const timeStampCurrenciesUpdated = useSelector(
    (state) => state.rates.updatedAt
  );
  const allCurrenciesNames = allCurrencies.map((el) => el.name);

  const dispatch = useDispatch();

  const [chosenCurrency, setChosenCurrency] = useState("USD");
  const [chosenRatesSource, setChosenRatesSource] = useState("MasterCard");

  const [keepAPIRatesCache, setkeepAPIRatesCache] = useState(true);

  function handleListChange(e) {
    const { name, value } = e.target;
    if (name === "currency") {
      setChosenCurrency((prev) => value);
      setValue(value, 1);
    }
    if (name === "ratesSource") {
      setChosenRatesSource((prev) => value);
    }
    return;
  }

  function updateRatesIfCacheExpired() {
    const timeStampNow = new Date();
    const timePassAfterRatesUpdated =
      (timeStampNow - timeStampCurrenciesUpdated) / 1000;

    if (timePassAfterRatesUpdated >= ratesUpdatingTimeFrame) {
      dispatch(getNewRatesThunkCreator());
    }

    if (!keepAPIRatesCache) {
      dispatch(getNewRatesThunkCreator());
    }
  }

  useEffect(() => {
    if (ratesSource === "Manual") {
      setkeepAPIRatesCache((prev) => false);
    }
    if (ratesSource === "MasterCard") {
      setkeepAPIRatesCache((prev) => true);
    }
  }, [ratesSource]);

  const onSubmit = ({
    price,
    time: timeString,
    currency,
    ratesSource,
    ...manualRates
  }) => {
    dispatch(setRequestErr(false));

    const time = convertStrTimeToNum(timeString);

    if (ratesSource === "MasterCard") {
      updateRatesIfCacheExpired();
    }

    if (ratesSource === "Manual") {
      dispatch(setManualRates(manualRates));
    }

    dispatch(submitFieldsData({ price, time, currency }));
  };

  useEffect(() => {
    setValue("UAH", "");
    setValue("RUB", "");
    setValue("EUR", "");
    setValue("USD", "");
    setValue(chosenCurrency, 1);
  }, [chosenCurrency]);

  return (
    <div
      className={
        state.darkMode ? "form-container" : "form-container light-form"
      }
    >
      <form id="calc-form" onSubmit={handleSubmit(onSubmit)}>
        <Input
          inputName="price"
          register={register}
          labelName={t("labelPrice")}
          placeholder={t("pricePlaceholder")}
          changeHandler={handlePriceChange}
          errors={errors}
        />
        <Input
          inputName="time"
          register={register}
          labelName={t("labelTime")}
          placeholder={t("timePlaceholder")}
          changeHandler={handleTimeChange}
          errors={errors}
        />
        <Select
          labelName={t("labelCurrency")}
          inputName="currency"
          register={register}
          changeHandler={handleListChange}
          value={chosenCurrency}
          optionsArr={allCurrenciesNames}
          errors={errors}
        />
        <Select
          labelName={t("labelExchangeRate")}
          inputName="ratesSource"
          register={register}
          changeHandler={handleListChange}
          value={chosenRatesSource}
          optionsArr={ratesSources}
          errors={errors}
        />
        {chosenRatesSource === "Manual" && (
          <RatesInputSet
            register={register}
            allCurrencies={allCurrencies}
            chosenCurrency={chosenCurrency}
            errors={errors}
          />
        )}
      </form>
    </div>
  );
}
