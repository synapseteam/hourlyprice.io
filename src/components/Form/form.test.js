/**
 * @jest-environment jsdom
 */

import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import { AppThemeProvider } from "context/AppContext";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";

import Form from "components/Form";

if (!global.React) {
  global.React = React;
}

const mockStore = configureStore();
const initState = {
  main: {
    fields: {
      price: "",
      time: "",
      currency: "USD",
    },
  },
  rates: {
    allCurrencies: [
      { name: "USD", rate: 1, symbol: "$", isoCode: "840" },
      { name: "EUR", rate: 1, symbol: "€", isoCode: "978" },
      { name: "UAH", rate: 1, symbol: "₴", isoCode: "980" },
      { name: "RUB", rate: 1, symbol: "₽", isoCode: "810" },
    ],
    updatedAt: "",
    ratesSource: "MasterCard",
  },
};

const context = {
  darkMode: true,
};
const value = [context];

const store = mockStore(initState);

test("Form component snapshot test", () => {
  const tree = renderer
    .create(
      <AppThemeProvider value={value}>
        <Provider store={store}>
          <Form />
        </Provider>
      </AppThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Form component: user submit empty form", async () => {
  const { container } = render(
    <AppThemeProvider value={value}>
      <Provider store={store}>
        <Form />
      </Provider>
    </AppThemeProvider>
  );

  const form = container.querySelector("form");
  const allLabels = form.querySelectorAll("label");

  const allLabelsTexts = [...allLabels].map((label) => label.textContent);

  expect(allLabelsTexts.length).toBe(4);

  await waitFor(() => {
    fireEvent.submit(form);
  });
  let allErrors = [...form.querySelectorAll(".input__error")];

  let allErrorsTexts = allErrors.map((err) => err.textContent);

  expect(allErrorsTexts.length).toBe(2);
});
