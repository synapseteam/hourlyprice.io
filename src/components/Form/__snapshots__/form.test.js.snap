// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component snapshot test 1`] = `
<div
  className="css-1v86ben-StylesContainer e6voiva1"
>
  <form
    className="css-1p4t8xb-StyledForm e6voiva0"
    id="calc-form"
    onSubmit={[Function]}
  >
    <label
      className="input__label css-rqm9rb-StyledLabel e1bg0i6e0"
    >
      Price
      :
      <input
        autoComplete="off"
        className="input css-kw93hu-StyledInput e1waaif10"
        inputMode="decimal"
        name="price"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="20.30"
        type="text"
      />
    </label>
    <label
      className="input__label css-rqm9rb-StyledLabel e1bg0i6e0"
    >
      Time
      :
      <input
        autoComplete="off"
        className="input css-kw93hu-StyledInput e1waaif10"
        inputMode="decimal"
        name="time"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="HH:MM"
        type="text"
      />
    </label>
    <label
      className="select__label css-rqm9rb-StyledLabel e1bg0i6e0"
    >
      Currency
      :
      <select
        className="select css-nn1g6a-StyledSelect e1q41qb90"
        name="currency"
        onBlur={[Function]}
        onChange={[Function]}
        value="USD"
      >
        <option
          className="select__option"
          value="USD"
        >
          USD
        </option>
        <option
          className="select__option"
          value="EUR"
        >
          EUR
        </option>
        <option
          className="select__option"
          value="UAH"
        >
          UAH
        </option>
        <option
          className="select__option"
          value="RUB"
        >
          RUB
        </option>
      </select>
    </label>
    <label
      className="select__label css-rqm9rb-StyledLabel e1bg0i6e0"
    >
      Exchange rate
      :
      <select
        className="select css-nn1g6a-StyledSelect e1q41qb90"
        name="ratesSource"
        onBlur={[Function]}
        onChange={[Function]}
        value="MasterCard"
      >
        <option
          className="select__option"
          value="MasterCard"
        >
          MasterCard
        </option>
        <option
          className="select__option"
          value="Manual"
        >
          Manual
        </option>
      </select>
    </label>
  </form>
</div>
`;
